### PR TITLE
Implement functions for minimum-sample-size-calculators

### PR DESCRIPTION
**Pull request by ghost**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
New functions minSampleSizeContinous and minSampleSizeConversion 

I Implemented two functions for minimum sample size calculators.

### `SELECT minSampleSizeContinous(20, 10, 0.05, 0.8, 0.05)`
  - Parameters
    - Mean of control-metric
    - Standard deviation of conrol-metric
    - Minimal Detectable Effect
    - Sufficient statistical power to detect a treatment effect
    - Significance level
  - Return
    - A tuple containing minimum_sample_size, detect_range_lower, and detect_range_upper.

### `SELECT minSampleSizeConversion(0.9, 0.01, 0.8, 0.05)`
  - Parameters
    - Mean of control-metric
    - Minimal Detectable Effect
    - Sufficient statistical power to detect a treatment effect
    - Significance level
  - Return
    - A tuple containing minimum_sample_size, detect_range_lower, and detect_range_upper.

These functions can be used in AB Test and others.

The descriptions of formulas are in [LINK](https://towardsdatascience.com/required-sample-size-for-a-b-testing-6f6608dd330a).

### CODE REVIEW

File: src/Functions/minSampleSize.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+            return std::make_shared<DataTypeTuple>(
+                std::move(types),
+                std::move(names)
+            );
+        }
+
+        DataTypePtr getReturnTypeImpl(const DataTypes & /*arguments*/) const override
+        {
+            return getReturnType();
```
Comment (kitaisreal) [inline src/Functions/minSampleSize.cpp:62]: Need to validate arguments types, and throw exception if arguments types mismatch.

File: src/Functions/minSampleSize.cpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        }
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+        {
+            MutableColumnPtr to{getReturnType()->createColumn()};
```
Comment (kitaisreal) [inline src/Functions/minSampleSize.cpp:67]: Current implementation is extremely slow.
First we can safely unwrap arguments to columns implementations, ColumnFloatUInt64, ColumnFloat64,  ColumnFloat64.
Second we can unwrap result tuple column in 3 columns UInt64, Float64, Float64, resize data to necessary rows count, and then just insert by index.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Functions/minSampleSize.cpp (hunk 1 above)
Comment (achimbab) [reply src/Functions/minSampleSize.cpp:76]: I added codes in order to validate arguments.

File: src/Functions/minSampleSize.cpp (hunk 2 above)
Comment (achimbab) [reply src/Functions/minSampleSize.cpp]: I unwrapped arguments and the result tuple.

File: src/Functions/minSampleSize.cpp [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+            return ColumnTuple::create(Columns{std::move(res_min_sample_size), std::move(res_detect_lower), std::move(res_detect_upper)});
+        }
+
+        static ColumnPtr execute(const ColumnsWithTypeAndName & arguments, size_t input_rows_count)
+        {
+            const auto & arg_baseline = arguments[0];
+            const auto & arg_sigma = arguments[1];
+            ColumnPtr res;
+
+            const bool valid = castBothTypes(arg_baseline.type.get(), arg_sigma.type.get(), [&](const auto & left, const auto & right)
```
Comment (kitaisreal) [inline src/Functions/minSampleSize.cpp:274]: @achimbab not sure, why this is needed here ? It seems that we have fixed types for these functions, or not ?

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Functions/minSampleSize.cpp (hunk 3 above)
Comment (achimbab) [reply src/Functions/minSampleSize.cpp]: @kitaisreal I removed this codes.

File: src/Functions/minSampleSize.cpp [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+            data_detect_lower.reserve(input_rows_count);
+
+            auto res_detect_upper = ColumnFloat64::create();
+            auto & data_detect_upper = res_detect_upper->getData();
+            data_detect_upper.reserve(input_rows_count);
+
+            for (size_t row_num = 0; row_num < input_rows_count; ++row_num)
+            {
+                /// Mean of control-metric
+                Float64 baseline = col_baseline.getFloat64(row_num);
```
Comment (kitaisreal) [inline src/Functions/minSampleSize.cpp:130]:
```c++
ColumnVector
Float64 getFloat64(size_t n) const override
{
    if constexpr (is_arithmetic_v<T>)
        return static_cast<Float64>(data[n]);
    else
        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Cannot get the value of {} as Float64", TypeName<T>);
}
```
it too heavy method, please use `getData` from Column to get array of Float64.

File: src/Functions/minSampleSize.cpp [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+                    data_detect_lower.emplace_back(std::numeric_limits<Float64>::quiet_NaN());
+                    data_detect_upper.emplace_back(std::numeric_limits<Float64>::quiet_NaN());
+                    continue;
+                }
+
+                Float64 delta = baseline * mde;
+
+                using namespace boost::math;
+                normal_distribution<> nd(0.0, 1.0);
+                Float64 min_sample_size = 2 * (std::pow(sigma, 2)) * std::pow(quantile(nd, 1.0 - alpha / 2) + quantile(nd, power), 2) / std::pow(delta, 2);
```
Comment (kitaisreal) [inline src/Functions/minSampleSize.cpp:152]: Please add original formulas and links into code.

File: src/Functions/minSampleSize.cpp [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+                    data_min_sample_size.emplace_back(std::numeric_limits<Float64>::quiet_NaN());
+                    data_detect_lower.emplace_back(std::numeric_limits<Float64>::quiet_NaN());
+                    data_detect_upper.emplace_back(std::numeric_limits<Float64>::quiet_NaN());
+                    continue;
+                }
+
+                Float64 delta = baseline * mde;
+
+                using namespace boost::math;
+                normal_distribution<> nd(0.0, 1.0);
```
Comment (kitaisreal) [inline src/Functions/minSampleSize.cpp:151]: It seems that distribution constructor should be moved out of hot loop.

File: src/Functions/minSampleSize.cpp [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+                    continue;
+                }
+
+                Float64 q1 = 1.0 - p1;
+                Float64 p2 = p1 + mde;
+                Float64 q2 = 1.0 - p2;
+                Float64 p_bar = (p1 + p2) / 2.0;
+
+                using namespace boost::math;
+                normal_distribution<> nd(0.0, 1.0);
```
Comment (kitaisreal) [inline src/Functions/minSampleSize.cpp:224]: Same as for `ContinousImpl`.

**Review by kitaisreal — APPROVED** [review]
Looks good. Please consider to make additional small improvements. Then it is ready to be merged.

**Review by kitaisreal — CHANGES_REQUESTED** [review]
Need to fix small comments.

File: src/Functions/minSampleSize.cpp (hunk 4 above)
Comment (achimbab) [reply src/Functions/minSampleSize.cpp]: I fixed it to use `getData`.

File: src/Functions/minSampleSize.cpp (hunk 5 above)
Comment (achimbab) [reply src/Functions/minSampleSize.cpp]: I added the fomula and a link.

File: src/Functions/minSampleSize.cpp (hunk 6 above)
Comment (achimbab) [reply src/Functions/minSampleSize.cpp]: I moved distribution constructor out of the hot loop.

File: src/Functions/minSampleSize.cpp (hunk 7 above)
Comment (achimbab) [reply src/Functions/minSampleSize.cpp]: OK. I fixed it.

**Review by kitaisreal — APPROVED** [review]
Will merge in merging branch to update documentation and check performance.

**Pull request merged**